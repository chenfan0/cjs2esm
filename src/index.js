import fs from 'node:fs'
import path from 'node:path'

import babel from '@babel/core'
import t from '@babel/types'
import traverse from '@babel/traverse'
import generate from '@babel/generator'

function isRequireCallExpression(node) {
  const isRequire = t.isIdentifier(node.callee, { name: 'require' }) && node.arguments.length === 1

  if (!isRequire) {
    return false
  }

  const arg = node.arguments[0]

  if (t.isStringLiteral(arg)) {
    return {
      type: 'static',
      requireId: arg.value
    }
  } 
  return {
    type: 'dynamic',
    requireId: arg.name
  }
  
}

function isStaticRequireCallExpression(node) {
  return t.isCallExpression(node) &&
    t.isIdentifier(node.callee, { name: 'require' }) &&
    node.arguments.length === 1 &&
    t.isStringLiteral(node.arguments[0])
}

function completeRequirePath(basePath, requirePath) {
  /**
   * eg:
   *  node:path -> node:path
   *  dayjs -> dayjs
   *  . -> ./index.js
   *  ./ -> ./index.js
   *  ../ -> ../index.js
   *  ./file -> ./file.js or ./file/index.js
   * 
   */
  if (requirePath === '.') {
    const exist = fs.existsSync(path.resolve(basePath, 'index.js'))
    let completedRequirePath = '.'
    if (exist) {
      completedRequirePath = './index.js'
    }
    return {
      completedRequirePath,
      isLocalModule: true
    }
  } else if (
    requirePath.startsWith('./') 
    || 
    requirePath.startsWith('../')
  ) {
    const ext = path.extname(requirePath)
    if (!ext) {
      const exist = fs.existsSync(path.resolve(basePath, requirePath + '.js'))
      let completedRequirePath = requirePath
      if (exist) {
        completedRequirePath = requirePath + '.js'
      } else {
        const existDir = fs.existsSync(path.resolve(basePath, requirePath, 'index.js'))
        if (existDir) {
          completedRequirePath = path.join(requirePath, 'index.js')
        }
      }
      return {
        completedRequirePath,
        isLocalModule: true
      }
    } else {
      return {
        completedRequirePath: requirePath,
        isLocalModule: true
      }
    }
  } else {
    return {
      completedRequirePath: requirePath,
      isLocalModule: false
    }
  }
}

function isDynamicRequireCallExpression(node) {
  return t.isCallExpression(node) &&
    t.isIdentifier(node.callee, { name: 'require' }) &&
    node.arguments.length === 1 &&
    t.isIdentifier(node.arguments[0])
}

function getRequireArg(node) {
  if (isStaticRequireCallExpression(node)) {
    return node.arguments[0].value
  }
  return node.arguments[0].name

}

export function transformCjsToEsm(filePath) {
  const code = fs.readFileSync(filePath, 'utf-8')

  const ast = babel.parse(code)
  const requirePaths = []

  traverse(ast, {
    VariableDeclaration(traversePath) {
      const declarations = traversePath.node.declarations

      if (declarations.length === 1) {
        const declaration = declarations[0]
        const id = declaration.id
        const init = declaration.init

        if (!isRequireCallExpression(init)) {
          return
        }

        const { type, requireId } = isRequireCallExpression(init)
        const { completedRequirePath, isLocalModule } = completeRequirePath(path.dirname(filePath), requireId)

        if (isLocalModule) {
          requirePaths.push(completedRequirePath)
        }

        if (type === 'static') {
          if (t.isIdentifier(id)) {
            /**
             * eg: 
             *  const fs = require("node:fs")
             */
            traversePath.replaceWith(
              t.importDeclaration(
                [t.importDefaultSpecifier(id)],
                t.stringLiteral(completedRequirePath)
              )
            )
          } else if (t.isObjectPattern(id)) {
            /**
             * eg:
             *  const { port: port1, name1 } = require('../../config')
             */
            const properties = id.properties
            const specifiers = properties.map(property => {
              return t.importSpecifier(property.value, property.key)
            })
            traversePath.replaceWith(
              t.importDeclaration(
                specifiers,
                t.stringLiteral(completedRequirePath)
              )
            )

          }

        } else {
          // 动态引入
        }


      } else {
        const needDelete = []
        for (let i = 0; i < declarations.length; i++) {
          const declaration = declarations[i]
          const id = declaration.id
          const init = declaration.init

          if (!init || !isRequireCallExpression(init)) {
            continue
          }

          const { type, requireId } = isRequireCallExpression(init)
          const { completedRequirePath, isLocalModule } = completeRequirePath(path.dirname(filePath), requireId)

          if (isLocalModule) {
            requirePaths.push(completedRequirePath)
          }

          if (type === 'static') {
            if (t.isIdentifier(id)) {
              traversePath.insertBefore(
                t.importDeclaration(
                  [t.importDefaultSpecifier(id)],
                  t.stringLiteral(completedRequirePath)
                )
              )
              needDelete.push(i)
            } else if (t.isObjectPattern(id)) {
              const properties = id.properties
              const specifiers = properties.map(property => {
                return t.importSpecifier(property.value, property.key)
              })
              traversePath.insertBefore(
                t.importDeclaration(
                  specifiers,
                  t.stringLiteral(completedRequirePath)
                )
              )
              needDelete.push(i)
            }
          } else {
            // 动态引入
          }

        }

        traversePath.node.declarations = declarations.filter((_, i) => !needDelete.includes(i))
      }
    },
  });

  const res = generate(ast).code;

  return res
}
