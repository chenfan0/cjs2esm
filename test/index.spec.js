import path from "node:path";
import fs from "node:fs";

import { describe, it, expect } from "vitest";
import { transformCjsToEsm } from "../src";

const dirname = import.meta.dirname;

describe("cjs -> esm", () => {
  it("static-one-identifier", () => {
    const cjsPath = path.resolve(dirname, "./cjs/static-one-identifier.js");
    expect(transformCjsToEsm(cjsPath)).matchSnapshot();
  });

  it("static-multi-identifier", () => {
    const cjsPath = path.resolve(dirname, "./cjs/static-multi-identifier.js");
    expect(transformCjsToEsm(cjsPath)).toMatchSnapshot();
  });

  it('static-object-pattern', () => {
    const cjsPath = path.resolve(dirname, './cjs/static-object-pattern.js')
    expect(transformCjsToEsm(cjsPath)).toMatchSnapshot()
  })

  it('static-require-path-complete', () => {
    const cjsPath = path.resolve(dirname, './cjs/static-require-path-complete.js')
    expect(transformCjsToEsm(cjsPath)).toMatchSnapshot()
  })

});
