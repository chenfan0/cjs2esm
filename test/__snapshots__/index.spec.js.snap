// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`cjs -> esm > static-multi-identifier 1`] = `
"import fs from "node:fs";
import { resolve, relative as rel } from "node:path";
import { test } from "./static-object-pattern.js";
let name;"
`;

exports[`cjs -> esm > static-object-pattern 1`] = `"import { readFile, readFileSync as rfs } from "node:fs";"`;

exports[`cjs -> esm > static-one-identifier 1`] = `"import fs from "node:fs";"`;

exports[`cjs -> esm > static-require-path-complete 1`] = `
"import path from "node:path";
import staticMultiIdentifier from "./static-multi-identifier.js";
import index from "../../src/index.js";"
`;
